File: Python/find-the-substring-with-maximum-cost.py
# Time:  O(n)
# Space: O(1)

import itertools


# greedy, kadane's algorithm
class Solution(object):
    def maximumCostSubstring(self, s, chars, vals):
        """
        :type s: str
        :type chars: str
        :type vals: List[int]
        :rtype: int
        """
        def kadane(s):
            result = curr = 0
            for c in s:
                curr = max(curr+(lookup[c] if c in lookup else ord(c)-ord('a')+1), 0)
                result = max(result, curr)
            return result

        lookup = {}
        for c, v in itertools.izip(chars, vals):
            lookup[c] = v
        return kadane(s)

File: C++/find-the-substring-with-maximum-cost.cpp
// Time:  O(n)
// Space: O(1)

// greedy, kadane's algorithm
class Solution {
public:
    int maximumCostSubstring(string s, string chars, vector<int>& vals) {
        unordered_map<char, int> lookup;
        for (int i = 0; i < size(chars); ++i) {
            lookup[chars[i]] = vals[i];
        }
        const auto& kadane = [&](const auto& s) {
            int result = 0, curr = 0;
            for (const auto& c : s) {
                curr = max(curr + (lookup.count(c) ? lookup[c] : c - 'a' + 1), 0);
                result = max(result, curr);
            }
            return result;
        };

        return kadane(s);
    }
};
